Feature/new architecture proposal

marcellodesales opened the pull request:
This pull request is to improve and solve on Issue #15 ... In order to properly work, empty files should be passed to the stream `gulp.src()`. The README and Test cases were updated to reflect this proposal.

API Change
---------

Just changing the plugin to use streams and properly handle different protocols:
* file://
* http://
* https://
* File paths starting with "/"

```js
var withFilePath = {
  uri: process.env.PWD + "/package.json"
};
gulp.src([]).pipe(open(withFilePath));

// Open from the stream and default program
gulp.src("./package.json").pipe(open());

// Open from the stream using the app
var chrome = os.platform() === 'linux' ? 'google-chrome' : (
    os.platform() === 'darwin' ? 'google chrome' : (
    os.platform() === 'win32' ? 'chrome' : 'firefox'));
  gulp.src("./package.json").pipe(open({app: chrome}));
```

Output
-------

The task will show if a file or a URL was opened and if that was with the default or with the given app.

```
[19:04:56] Starting 'open'...
[19:04:56] Finished 'open' after 1.25 ms
[19:05:44] Opening the url http://localhost:3001/?ws=localhost:3001&port=5858 using the default OS app
[19:04:56] Opening the file /home/mdesales/dev/github-intuit/servicesplatform-nodejs/sp-reference-restify/package.json using the default OS app
```

Categorized Test Cases
------

```
$ npm test

> gulp-open@0.3.2 test /home/mdesales/dev/github/gulp-open
> mocha --reporter spec

  gulp-open
    opening files
      ✓ should open a stream of files with the default application
      ✓ should open a stream of files with a defined application
      ✓ should open a stream of files with a browser using the options
    opening urls
      ✓ should open a website in a browser using the options
      ✓ should open a website in a browser using chrome or firefox

  5 passing (5ms)
```

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+var es = require('event-stream');
```
Use through2 as it is the gulp standard. Don't use es

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+var es = require('event-stream');
+var fs = require('fs');
```
Do not use any `fs` calls in any plugin

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
-      open(file.path);
-      return cb(null, file);
+// The acceptable uri formats for opening
+var ACCEPTABLE_URI_FORMATS = ["/", "file://", "http://", "https://"];
```
use single quotes

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+  // Executes the opening based on a given uri
+  var executeOpening = function() {
+    if (!opt.uri) {
+      stream.emit("error", new PluginError(PLUGIN_NAME, "URI is missing or incorrect! Use the src or the " + 
```
single quotes

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+    }
+
+    // If the uri was not passed either by the stream nor by the options
+    var indexOfIndexOf = _indexOfIndexOf(opt.uri);
```
these names are odd: `indexOfIndexOf`,  `_indexOfIndexOf`

marcellodesales replied:
Definitely :) Let me find something better....

marcellodesales replied:
Is the following better?

```js
    // If the uri was not passed either by the stream nor by the options
    var indexOfAnyFormat = _getIndexesOfAcceptableFormats(opt.uri);
    if (indexOfAnyFormat === -1) {
      stream.emit('error', new PluginError(PLUGIN_NAME, 'URI is missing or incorrect! Please use one of ' + ACCEPTABLE_URI_FORMATS));
    }

    // If the type is the index value in the range of files
    var selectedType = indexOfAnyFormat <= 1 ? 'file' : 'url';
```

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+
+    // If the type is the index value in the range of files
+    var selectedType = indexOfIndexOf <= 1 ? "file" : "url";
+    console.log(indexOfIndexOf);
```
do not use any `console.log`

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+      try {
+        fs.statSync(opt.uri);
+
+      } catch (FileNotFoundError) {
```
do not use full camel case

marcellodesales replied:
Block removed as I will not use "fs" to verify if the file exists.

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
-      return cb(null, file);
+  };
+
+  function done() {
```
this whole 'done' thing is fixed in through2

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+
+  // Handles opening the file through the stream
+  var queueFile = function(file) {
+    console.log(file);
```
no logs

marcellodesales replied:
Yeah completely missed that :(

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+  };
+
+  // Open the object, whatever that is.
+  var endStream = function() {
```
don't wrap unneeded functions

stevelacy commented on index.js:
```diff
@@ -1,25 +1,104 @@
+  };
+
+  // copied from gulp-karma
+  stream = es.through(queueFile, endStream);
```
use through2.obj, not es.

stevelacy commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "gulp-open",
   "description": "Open files and URLs with gulp",
-  "version": "0.3.2",
+  "version": "1.0.0",
```
don't bump the version

marcellodesales replied:
Will revert that..

stevelacy commented on test/main.js:
```diff
@@ -1,19 +1,35 @@
 var gulpopen = require('../');
 var should = require('should');
+var os = require("os");
```
use single quotes

stevelacy commented:
Thanks for spending time writing this.
I commented on a few issues I noticed.

I forgot to add the correct linting file to the project as found here: https://github.com/stevelacy/gulp-stylus/blob/master/.jshintrc

There seems to be quite a bit of code added which could be simplified, first bt using through2, and second by having smaller functions for looping, etc.
There are some checks which are handled by gulp, if not then `open` handles it.

marcellodesales commented:
@stevelacy Thanks for doing code review... Yeah, I did not use lint for the lack of it... Also, I did not know about the use of through2 vs es... I've seen this pattern in other modules before and I will have to read the docs from both in order to learn and use them properly...

So, before I spend any time doing this, do you think you would be able to merge the pull request in case I address all the required changes?

thanks a lot!
Marcello

stevelacy commented:
Yes, I will merge it with the changes, provided that the examples work. They most likely will need to be updated to match your new API.

As a simple example of a through2 implementation: https://github.com/stevelacy/gulp-stylus/blob/master/index.js

emits are replaced with a simple return `cb(Error, Data)` pattern.
 
Thanks

marcellodesales commented:
@stevelacy Here's the first draft of the change... Can you please take a look again? 

I wrote this without seeing your message here :( I just reverted what you had done before and provided the updated samples in the README file, which was the original suggestion. 

Let me know if there's anything else I can do! Thanks a lot for the pointers!

stevelacy commented:
Thanks for making those changes, I commented on two two items I noticed if you don't mind modifying, I'll then merge it.

marcellodesales commented:
Hi @stevelacy... 

Thanks for another code review :+1: Yeah, I did not know what you would prefer, so I added the wget :) Just removed it... I also removed the reference to the array, and I'm using the suggested way for cb on errors.
